botblox/botblox-manager-software
# feat: Add support for tagged VLANs configuration

This PR adds support for configuring tagged VLAN on Switchblox.

It comes with a larger architectural change as the tagged VLAN is more complicated and the dict-based approach was not sufficient (as e.g. `--vlan-mode` is one setting that can change 2-3 fields depending on its value).

I created subpackage `switch` which contains the description of the hardware, its registers and data fields. Package `data_manager` contains methods that translate the CLI args to a common semantic representation (that should be switch-agnostic), and also methods for converting this common representation to the values of registers of a particular switch chip.

I also added the option to pass `test` as the serial device, in which case no serial communication is attempted. I'm not sure how were the `run_command_to_error` tests working before, but if they did not try to open the serial device, then it was by an error and not by intention.

I left the existing commands (`vlan` and `mirror`) and their tests with as few changes as possible. However, I fixed two wrong tests (`test_set_rx_mode` and `test_set_tx_mode`) where there was a strange combination of using `run_command_to_error` used for checking legitimate commands. There were also some tests that passed `self.base_args` twice to `run_command_to_error`, which caused the commands to fail on unknown command options. If you like the design of the changes this PR brings in, both existing commands can be transformed to use it, too (going against #14, though).

I also added the `erase` command.

The CLI parser now adjusts the available options to the switch chip the command uses. There is argument `--switch` with default value `switchblox`, which is passed to the subparsers and they can decide which options are valid for the particular chip. In future, the manager could attempt to automatically detect the chip and adjust the commands automagically.

I did not use the `mutually_exclusive_group` for `--reset` as I found out that it blocks displaying the full help message, which is really essential with this complex command.

## To test tagged VLAN

To test that VLANs are working I use the following setup: configure switch as in the `tagvlan/test_combination` test:

    botblox -D /dev/ttyUSB0 tag-vlan --port-default-vlan 1 2 --port-default-vlan 4 2 --vlan-mode OPTIONAL --port-header-action 1 ADD --port-header-action 4 STRIP --vlan 2 1 4

This tells the switch to label all packets from ports 1 and 4 as VLAN ID 2. Further, all packets going out of port 4 have their VLAN tag header removed, and all packets going out of port 1 have the VLAN tag header added.

Connect computer C1 to port 4, and computer C2 to port 1. Set C1 static address:

    ip addr add 192.168.2.150/24 dev eth0

And instruct C1 to ping C2 (even before C2 is set up):

    ping 192.168.2.1

Now open a console on C2 and set up a tagged VLAN interface:

    sudo apt install vlan
    sudo modprobe 8021q
    ip link add link eth0 name eth0.2 type vlan id 2
    ip addr add 192.168.2.1/24 brd 192.168.2.255 dev eth0.2
    ip link set dev eth0.2 up
    # or you can use nm-connection-editor to set up the VLAN interface and then `nmcli con up` to activate it

C1 should now start getting the pings back.

Open Wireshark on C2 on interface `eth0` (not on `eth0.2`!). You will see the ping ICMP messages and some other traffic. If you inspect the packets, they should have the 802.1Q layer:

![Snímek z 2021-04-27 03-55-00](https://user-images.githubusercontent.com/182533/116173184-6b811580-a70c-11eb-9931-7dacd629ba5d.png)

If you open Wireshark on C1's `eth0`, no packet should have the 802.1Q layer.
